### PR TITLE
Fix Client Websocket

### PR DESCRIPTION
**Pull request by MJHC**
Refactored websocket on client

### CODE REVIEW

File: client/src/socket_manager.rs [hunk 1]
```diff
@@ -0,0 +1,577 @@
+use futures_util::lock::Mutex;
+use futures_util::{Sink, SinkExt, Stream, StreamExt};
+use tokio::net::TcpStream;
+use tokio_tungstenite::tungstenite::client::IntoClientRequest;
+use tokio_tungstenite::tungstenite::protocol::{CloseFrame, WebSocketConfig};
+use tokio_tungstenite::{
+    client_async_tls_with_config, tungstenite, Connector, MaybeTlsStream, WebSocketStream,
+};
+
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:9]: Remove

File: client/src/socket_manager.rs [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+use rustls_pemfile::certs;
+use socket2::{SockRef, TcpKeepalive};
+use std::fs::File;
+use std::io::BufReader;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::broadcast::{self, Receiver, Sender};
+use tokio_tungstenite::tungstenite::Message;
+
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:23]: Remove

File: client/src/socket_manager.rs [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        std::pin::Pin::new(&mut self.0).poll_ready(cx)
+    }
+
+    fn start_send(mut self: std::pin::Pin<&mut Self>, item: Message) -> Result<(), Self::Error> {
+        std::pin::Pin::new(&mut self.0).start_send(item)
+    }
+
+    fn poll_flush(
+        mut self: std::pin::Pin<&mut Self>,
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:111]: Please use imports

File: client/src/main.rs [hunk 4]
```diff
@@ -1,22 +1,75 @@
...
+    });
+    println!("---------------------------------------------");
+    let req = create_request(
+        id,
+        "PUT",
+        "/v1/messages/7db772c1-5ae2-4d25-9daf-025be34aa7b1",
+        vec![],
+        Some(body),
+    );
+    println!("{:?}", socket_manager.send(id, req).await.expect("benis"));
```
Comment (Darkros1245) [inline client/src/main.rs:71]: Hmm, Benis?

File: client/src/socket_manager.rs [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    let connector = Connector::Rustls(Arc::new(tls_cfg));
+
+    let config = WebSocketConfig {
+        max_frame_size: Some(0x210000),
+        ..Default::default()
+    };
+
+    let res = client_async_tls_with_config(req, stream, Some(config), Some(connector)).await;
+    let (ws, _) = res.map_err(|_| "Failed to connect to server".to_string())?;
+    Ok(ws)
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:187]: Why not:
```
    res.map_err(|_| "Failed to connect to server".to_string())
```

File: client/src/socket_manager.rs [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+                        break;
+                    },
+                };
+
+                tokio::time::sleep(Duration::from_millis(KEEPALIVE_INTERVAL_MS.into())).await;
+            }
+        });*/
+
+        Ok(())
+    }
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:320]: Does this need to be a part of this PR or should this just be todo?

File: client/src/socket_manager.rs [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+
+    async fn close(&mut self) {
+        let mut guard = self.connection.lock().await;
+        if let ConnectionState::Active(mut socket) =
+            std::mem::replace(&mut *guard, ConnectionState::Closed)
+        {
+            let _ = socket
+                .send(Message::Close(Some(CloseFrame {
+                    code: tungstenite::protocol::frame::coding::CloseCode::Normal,
```
Comment (Mikkel-Boje) [inline client/src/socket_manager.rs:329]: PLEASE IMPORT

File: common/src/websocket/connection_state.rs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+use futures_util::stream::SplitSink;
+use std::fmt::Debug;
+
+#[derive(Debug)]
+pub enum ConnectionState<M, T> {
+    Active(SplitSink<T, M>),
```
Comment (Darkros1245) [inline common/src/websocket/connection_state.rs:6]: T and M needs to be in the same order

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]

File: server/src/server.rs [hunk 9]
```diff
@@ -93,6 +83,7 @@ pub async fn handle_put_messages<T: SignalDatabase, U: WSStream + Debug>(
         .iter()
         .map(|message| message.destination_device_id)
         .collect();
+    println!("here");
```
Comment (Darkros1245) [inline server/src/server.rs:86]: Remove

File: client/src/socket_manager.rs [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+const KEEPALIVE_INTERVAL_MS: u32 = 30 * SECOND;
+
+// If the machine was in suspended mode for more than 5 minutes - trigger
+// immediate disconnect.
+const STALE_THRESHOLD_MS: u32 = 5 * MINUTE;
+
+// If we don't receive a response to keepalive request within 30 seconds -
+// close the socket.
+const KEEPALIVE_TIMEOUT_MS: u32 = 30 * SECOND;
+
```
Comment (Darkros1245) [inline client/src/socket_manager.rs:44]:
```suggestion
```

File: client/src/socket_manager.rs [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+                    Err(_) => {
+                        mgr.close_reason(3008.into(), "Timed out".into()).await;
+                        break;
+                    },
+                };
+
+                tokio::time::sleep(Duration::from_millis(KEEPALIVE_INTERVAL_MS.into())).await;
+            }
+        });*/
```
Comment (Darkros1245) [inline client/src/socket_manager.rs:317]:
```suggestion
```

File: client/src/socket_manager.rs [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+                        mgr.close().await;
+                        break;
+                    }
+                    _ => {
+                        continue;
+                    }
+                }
+            }
+        });
+        let mgr = self.clone();
```
Comment (Darkros1245) [inline client/src/socket_manager.rs:274]:
```suggestion
```

**Review by Darkros1245 — CHANGES_REQUESTED** [review]

**Review by Darkros1245 — DISMISSED** [review]

**Review by Mikkel-Boje — DISMISSED** [review]
:+1:

**Review by Mikkel-Boje — APPROVED** [review]

**Review by Darkros1245 — APPROVED** [review]

**Pull request merged**